### DISCUSSION TITLE
Advice / suggestion on how to implement 'integration tests simulating loss-of-power' under JUnit5 (migration from JUnit4) ?

### DISCUSSION THREAD

**Discussion by oliva123456** [junit-team/junit-framework#3098] [Q&A]
Hi,

we are currently migrating from JUnit 4 and i'm quite unsure how to migrate a special test scenario.

aside from simple junit tests we additionally have integration tests: we start up our application and connect the test class to our application (e.g. user interaction). we were already able to migrate integration tests to JUnit5.

but we have some special integration tests, in which we simulate a loss of power within our application:
- there is a phase before a loss of power: first instance of our application is started, used (e.g. with user input) and terminated (when the test method ends)
- there is a phase after a loss of power: second instance of our application is started and the behaviour of the application is checked
- both phases run in the JUnit lifecycle (within JUnit4: Rules/ClassRules/Assertions)
- 'the application' is started in the same JVM as the test

now i'm quite unsure, how to implement the scenario within JUnit5:
- implement our own TestEngine from scratch (this should work, but means re-implementing a lot of stuff from the existing test engines)
- implement our own TestEngine by subclassing org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine (after experimenting a bit imho the differentiation between before and after loss-of-power-phase does not fit well into the concept of hierarchical tests, maybe i missed something)
- using org.junit.jupiter.engine.JupiterTestEngine with some features, i haven't discovered so far
- using some other features like DynamicTest / TestFactory (i did not take a deep dive so far) 
- some other features 

**i'd appreciate very much, if someone already had similar scenarios or is able to give some advice ;-)**

for the sake of completeness: in JUnit4 we implemented those tests in the following way:
- a single test for each scenario consisting of:
	exactly one method annotated with our own marker-annotation @BeforeRecovery (running the phase before a loss of power)
	exactly one method annotated with @Test (running the phase after a loss of power)
- we subclassed our own TestRunner (OurTestRunner extends BlockJUnit4ClassRunner)
- we hooked into some methods of BlockJUnit4ClassRunner
   e.g.: 
   #computeTestMethods to handle the method @BeforeRecovery over to JUnit 
   #classBlock to run the before-recovery-phase before the proper test-method is started (another instance of OurTestRunner is started with a boolean, which indicates before/after loss-of-power).
   #methodBlock/#methodInvoker for the lifecycle
- the implementation probably is not ideal, but it works quite well.

for a better understanding some code of such a test:

```
  @RunWith(OurTestRunner.class)
  public class RecoveryTest {

      @BeforeClass
      public static void configureSomething() {
	  // configuration for both phases.
      }
  
      /**
       * Pre-recovery phase - run your testcase until the loss-of-power should happen.
       */
      @BeforeRecovery
      public void prepareForRecovery() {
          assertThat(getDisplay(), displays("something"));
          getDisplay().touch("something");
          ....
      }
  
      /**
       * After-recovery phase. Our application should restore the transaction
       * and the testcase might perform all checks.
       */
      @Test public void postRecovery() throws Exception {
          assertThat("something");
      }
  }

```  

thanks in advance,
Oliver

**Comment by portyu9** [comment]
Hi @oliva123456 

I would **not** implement a custom `TestEngine` for this. The scenario can be modeled as one Jupiter test invocation with an extension orchestrating the two application lifetimes.

A good fit is `InvocationInterceptor`.

For each recovery test:

1. start application instance no.1
2. invoke the method annotated with your custom `@BeforeRecovery`
3. simulate the power loss / stop instance no.1
4. start application instance no.2
5. let the normal `@Test` method execute
6. clean up instance no.2

For example, the important part of the extension can look like:

```java
class RecoveryExtension implements InvocationInterceptor {

    @Override
    public void interceptTestMethod(
            Invocation<Void> invocation,
            ReflectiveInvocationContext<Method> invocationContext,
            ExtensionContext context) throws Throwable {

        Object testInstance = context.getRequiredTestInstance();

        Method beforeRecovery = /* locate and validate the single
                                   @BeforeRecovery method */;

        startFirstApplication();

        try {
            context.getExecutableInvoker()
                    .invoke(beforeRecovery, testInstance);
        }
        finally {
            simulatePowerLossAndStopFirstApplication();
        }

        startSecondApplication();

        try {
            invocation.proceed(); // executes the regular @Test method
        }
        finally {
            stopSecondApplication();
        }
    }
}
```

and register it with:

```java
@ExtendWith(RecoveryExtension.class)
class RecoveryTest {
    ...
}
```

Using `context.getExecutableInvoker()` instead of plain reflection is useful because JUnit can still resolve supported method parameters.

This keeps both phases part of the same logical test, so a failed assertion in the pre-recovery phase also fails that test.

I would not use `DynamicTest`/`@TestFactory` simply to reproduce the two phases. Dynamic tests have a different lifecycle: individual dynamic tests do not receive the normal per-test lifecycle callbacks.

A custom `TestEngine` would only become worthwhile if the **pre-recovery and post-recovery phases themselves must be independently discovered and reported as separate test nodes** with custom engine semantics.

Please mark this answer as accepted if it helped, thank you.

**Comment by oliva123456** [comment]
Hey @portyu9,

thanks for your answer. 

In the meanwhile we did some prototyping with a result, which is at least a bit similar to your solution, but differs in details.
E.g.: 
* the pre-recovery-phase is run via JUnit's ConsoleLauncher to get an isolated process for this phase.
* InvocationInterceptor is a bit to late, as some initialisation code has already been performed at this point in the lifecycle

and some more.

lg,
Oliver

**Reply by portyu9** [reply]
Hey @oliva123456 

Thanks for confirming for us. Please mark my answer as accepted if it helped, thank you.
